src/components/multi-select/multi-select.test.jsx

```diff
@@ -0,0 +1,516 @@
+    expect(ref.current).toMatchSnapshot();
+  });
+
+  describe('classNames API', () => {
+    it('uses class name overrides provided', () => {
+      render(<MultiSelect
+        {...requiredProps}
+        classNames={{
+          container: 'unique-container',
+          formControlChildrenContainer: 'unique-form-control-children-container',
+          input: 'unique-input-class',
+          tagList: 'unique-tag-list-class',
+        }}
+      />);
+
+      expect(document.querySelector('.unique-container')).toBeInTheDocument();
+      expect(document.querySelector('.unique-form-control-children-container')).toBeInTheDocument();
+      expect(document.querySelector('.unique-input-class')).toBeInTheDocument();
+      expect(document.querySelector('.unique-tag-list-class')).toBeInTheDocument();
+    });
+  });
+
+  describe('id API', () => {
+    it('sets id of container and selected options', async () => {
+      render(<MultiSelect {...requiredProps} id="unique-id" value={['Foo']} />);
+
+      expect(document.querySelector('#unique-id')).toBeInTheDocument();
+      expect((await findSelectedOption('test label', 'Foo')).parentElement).toHaveAttribute('id', 'unique-id-option-Foo');
+    });
+  });
+
+  describe('label API', () => {
+    it('sets label of form control and clear button', async () => {
+      render(<MultiSelect {...requiredProps} label="unique label" value={['Foo']} />);
+
+      expect(screen.getByText('unique label')).toBeInTheDocument();
+      expect(await findRemoveButton('unique label', 'Foo')).toBeInTheDocument();
+    });
+  });
+
+  describe('listboxChildren API', () => {
+    it('uses render function for listbox children if set', async () => {
+      function listboxChildren(options, refs, onSelect) {
+        return (
+          <div>
+            {options.map((option, index) => (
+              <span key={option} ref={refs[index]} onClick={() => { onSelect({ target: refs[index] }); }}>Override {option}</span>
+            ))}
+          </div>
+        );
+      }
+
+      render(<MultiSelect {...requiredProps} listboxChildren={listboxChildren} />);
+
+      await openOptions('test label');
+      expect(screen.getByText('Override Foo')).toBeInTheDocument();
+      userEvent.click(screen.getByText('Override Foo'));
+      expect(await findSelectedOption('test label', 'Foo')).toBeInTheDocument();
+    });
+  });
+
+  describe('onChange API', () => {
+    it('fires the onChange event when the value changes', async () => {
+      let value = [];
+
+      function onChange(event) {
+        value = event.target.value;
+      }
+
+      render(<MultiSelect {...requiredProps} onChange={onChange} />);
+
+      await openOptions('test label');
+      userEvent.click(await await findAvailableOption('test label', 'Foo'));
+      expect(value).toContain('Foo');
+    });
+  });
+
+  describe('onInput API', () => {
+    it('fires the onInput event when the user types in the combobox input', async () => {
+      let value = null;
```

Comment:
This is odd. The standard patterns for testing callbacks is to use spies.

e.g. 

```
const spy = jest.fn();
render(<Component onInput={spy} />)
userEvent...
expect(spy).toHaveBeenCalledWith({ target: { value: ... }});
```

Reply:
I know we've used the pattern as written before, but I can use the spy pattern instead.

Reply:
This is actually not super straightforward, because of React `SyntheticEvents`. Basically, `toHaveBeenCalledWith` is going to be checking a `SyntheticEvent` that has already been returned to the pool, and therefore doesn't look like `{ target: { refStuff }}` anymore.

We can ensure that it was _called_ with this approach, but not what it was called with, unless we use `event.persist` which has other implications for production. See this comment from Kent Dodds, and the following 2: https://github.com/testing-library/react-testing-library/issues/175#issuecomment-422189880

For now, I'm going to go with `toHaveBeenCalled`. If we want to guarantee the event contract looks like what we need, I suggest we switch back to my non-standard pattern.